### ISSUE TITLE
Logger not recording IMU data

### ISSUE THREAD

**Issue by jvoermans** [jerabaul29/LoggerWavesInIce#3]
@jerabaul29 The logger is not logging any data, only generating empty files. This is what I did, perhaps you can see something I'm missing:

- I disconnected the RaspPi and Iridium modem from the logger.
- I uploaded the sketch "InitializeEEPROM.ino" on the ArduinoMega, waited for a while to let it run
- I uploaded the sketch "Logger_GPS_SD_VN.ino" on the ArduinoMega
- I configured the VN100 using the new Control Center of VectorNAV using following configuration:
-- System>Baud Rate: 57600 (I didn't specify Serial Port here)
-- Async Output Type>Ador: OFF (there is no "no output" here, also Serial Port is kept undefined here)
-- Async Output Freq>Adof: is kept at 40Hz
-- Binary Output Reg. #1> Port: Serial 1
-- Binary Output Reg. #1> Rate Divisor: 80
-- Binary Output Reg. #1>Message>IMU: UnCompMag, UnCompAccel, UnCompGyro, Temperature, Pressure
-- Binary Output Reg. #1>Message>Attitude: Ypr, Dcm, MagNed, AccNed

This gives a RATE of 10Hz, Size of 124 bytes (that is 1 byte more than the 123 of yours), BANDWITH 11900 bps and MESSAGE: $VNWRG,75,1,80,14,003E,003A*6ED3

-- IMU>Imu Filter Control: Mag, Accel, Gyro, Temp, Pres Window Size all set at 80.

Configuration is ended with: Write Settings>Execute

Disconnect and reconnect the IMU shows that all settings were configured.

One comment, the sketch says that the IMU has Serial2, is that something else than the Serial1 mentioned for the settings above? I did try Serial 2 in the settings as well, but didn't change anything with the Arduino output files.

There is also an option to load a configuration, we might try that as well if you have such a file.

If I connect the logger to my computer, the gps blinks normally, but the SD card blinks every 2 seconds with red light.

This is the serial output:

_Booting_

_Time1: 5000
Time2: 5000
card initialized.
GPS initialized
Init VN100 on Serial 2_

_----- End setup -----_

_str_rank: 19
File name: F00019
Start post...
Time log beginning: 6497
post success
Mess,Booting from beginning!
Time log end: 6525
Time log delay: 28
Start post...
Time log beginning: 6528
post success
R,732,CV,5.36
Time log end: 6536
Time log delay: 8_

_Digital in voltage: 732
Battery voltage (no calib.): 5.36
Start post...
Time log beginning: 6636
post success
$GPGGA,235957.800,,,,,0,00,,,M,,M,,*7F
M,6636
Time log end: 6642
Time log delay: 6
Start post...
Time log beginning: 6645
post success
R,575,CV,4.21
Time log end: 6653
Time log delay: 8_

_Digital in voltage: 575
Battery voltage (no calib.): 4.21_

And a little later, the GPS is giving time I guess?

_Start post...
Time log beginning: 151245
post success
$GPRMC,000638.312,V,,,,,0.00,0.00,231019,,,N*48
M,151245
Time log end: 151254
Time log delay: 9
Start post...
Time log beginning: 151258
post success
R,490,CV,3.59
Time log end: 151266
Time log delay: 8_

_Digital in voltage: 490
Battery voltage (no calib.): 3.59_

**Comment by jerabaul29** [comment]
Ok, so you uploaded the code for the logger on the instrument board, right? :) Are you using the instrument PCB with the logger code then?

A few things:

- the logger needs the IMU to be in the exact state I mention in the README (it looks like this is not the case currently? For example, the IMU filtering is not set), as it searches for the binary header to check the start of a valid message:

https://github.com/jerabaul29/LoggerWavesInIce/blob/bdcb0b8b306b9d6bfaa67e7240d75cab6b64ff0d/Logger_GPS_SD_VN_Binary_output/Logger_GPS_SD_VN_Binary_output.ino#L110-L115

So if the VN100 is in a different configuration, the binary header will be different, in which case the messages will not be recognized and the logger will just reboot. You may have a way to check for the binary header in the VN100 software. May you do that and check that your header is well ```fa 14 3e 00 3a 00```. If you want to use a different configuration, with a different header, you will need to change the header definition in the Arduino sketch.

**My recommendation here** is to use the **old** version of the VN100 software (still available on their website), and parameter the VN100 IMU **exactly** as is said in the readme.

- The configuration to use is slightly different between the waves in ice logger and the waves in ice instrument. Indeed, on the PCB of the waves in ice instrument, the VN100 is on the serial port Serial3. So you **do need (necessary condition)** to set the value there:

https://github.com/jerabaul29/LoggerWavesInIce/blob/bdcb0b8b306b9d6bfaa67e7240d75cab6b64ff0d/Logger_GPS_SD_VN_Binary_output/Logger_GPS_SD_VN_Binary_output.ino#L105

to:

```#define SERIAL_IMU Serial3```

recompile and upload.

- Did you upload the sketch using an extended buffer core for the Arduino Mega? See this:

Upload the right logger code on the Arduino Mega, using an **extended buffer core** (see ArduinoMegaExtendedSerialBuffers for more details). 

In the master readme of the repo.

This is also **necessary** for the good functionning of the logger. If you have difficulties using an extended buffer configuration, let me know I will help you. You can first try to either follow the explanations in the corresponding folder, or see on the internet (as the Arduino code architecture may have changed a bit). For example, platformio (if you use visual studio code + the platformio extension, which can be a good idea, there is a simple file where you can change such parameters. On my version of platformio, you can just update the platformio.ini of the project, and put something like:

build_flags = -D SERIAL_RX_BUFFER_SIZE=512 -D SERIAL_TX_BUFFER_SIZE=128

**Comment by jerabaul29** [comment]
If anything is unclear, just let me know. When this is sorted out I will update the readmes etc.

**Comment by jvoermans** [comment]
Sorry, the terminology is confusing me. Could you rephrase this for me: Ok, so you uploaded the code for the logger on the instrument board, right? :) Are you using the instrument PCB with the logger code then? I detached the arduino from the pcb, and uploaded both ino files on there. Or do I need to do specific handling to do that?

I tried to look for the old software but couldn't find it! Do you have a link for me?

Also confused about Serial3, I don't think in the VectorNav software there was an option for that, only 1 or 2.

I'll give it a go tonight or tmrw morning. If it fails, would you be able to do a skype session to walk me through it?

**Comment by jerabaul29** [comment]
I just mean that you use one of the PCBs you assembled, which are 'instruments' PCBs, but you want to use them to run the code of a logger, right? :) I.e. what we discussed last time :) .

- Old version of the VN100 software: here https://www.vectornav.com/support/downloads you can download the 'VectorNav Legacy Libraries' software
```
    VectorNav Sensor Explorer (Windows)
    EXE, 24.64 MB
```

- set to Serial3: this is to be done on the Mega code, not on the VN100. This is because the Arduino Mega has 3 serial ports, and on the PCB of the instrument the VN100 is on Port 3, while the current Arduino sketch for the logger expects Port 2

- I think I remember there is a tool for this somewhere in the VN100 software that allows to compute binary headers, but not sure where. But if you use the old VN100 software and the exact set of parameters I provide, should work.

- Yes, we could organize something like that.

**Comment by jerabaul29** [comment]
This is working now, right

For the record this was due to:

- not having programmed the VN100 with the right settings
- not using the right serial port
- not using the extended core

Fixing all 3 points gave good results (dropout 0.4% if I remember well?).

**Comment by jvoermans** [comment]
Yes. I tried first a readout of the additional sensors at 1Hz, directly
after the GPS code, but dropouts were 10%, so I just read them out now once
for every record together with voltage readout.

**Comment by jerabaul29** [comment]
Ok, strange. There must be a way to do it cleanly and get at least a few Hz of logging.
